### DISCUSSION TITLE
z.ZodSchema Questions

### DISCUSSION THREAD

**Discussion by nahtnam** [colinhacks/zod#794] [Q&A] [answered]
Hello,

I was looking for a way to define a human-readable type and then come up with the equivalent schema. This is what I ended up with:

```
type Dashboard = {
  id: string;
  type: string;
};
type Dashboards = {
  dashboards?: Dashboard[];
  currentDashboardId?: string;
  asdf?: string;
};

const dashboard: z.ZodSchema<Dashboard> = z.object({
  id: z.string(),
  type: z.string(),
});

export const dashboards: z.ZodSchema<Dashboards> = z.object({
  dashboards: z.array(dashboard).default([]),
  currentDashboardId: z.string().optional(),
});
```

Here are my two questions:

1. I noticed that here: `dashboards?: Dashboard[];` I have to put a `?` because there is `default` option provided. Is this expected behavior? I understand if it is but then it means that `.default` and `.optional` are similar and maybe the docs should note that? Any alternate solutions?
2. I noticed that any optional values such as `asdf` are not validated and typescript does not complain that it's not implemented

Any help would be appreciated. Thanks!

NOTE: Using `infer` works perfectly. I'd just prefer to have a human readable type visible to make sure that the types specified are correct.

**Comment by JacobWeisenburger** [comment, marked as answer]
> I noticed that here: dashboards?: Dashboard[]; I have to put a ? because there is default option provided. Is this expected behavior? I understand if it is but then it means that .default and .optional are similar and maybe the docs should note that? Any alternate solutions?

If you have a required field, then giving it a default value doesn't make any sense. Because, by definition, a required field always needs to be passed into `parse` or it will throw an error. So defining a default value would only ever make sense if that field was optional. So the docs could be updated to be more explicit about this, but I don't think it is entirely necessary. 

___
> I noticed that any optional values such as `asdf` are not validated and typescript does not complain that it's not implemented

- This isn't a question, this is a statement. :D :P
- Because `asdf` is an optional property of `Dashboards`, then `z.ZodSchema<Dashboards>` doesn't care if it's implemented or not because technically it isn't required by `Dashboards`. So the resulting type (`z.infer<typeof dashboards>`) is still compatible with `Dashboards`. If you define `asdf` on the schema with an invalid value type, then typescript complains.

```ts
export const dashboards: z.ZodSchema<Dashboards> = z.object( {
//           ^^^^^^^^^^
// The types of '_type.asdf' are incompatible between these types.
// Type 'number' is not assignable to type 'string | undefined'.
    asdf: z.number(),
    dashboards: z.array( dashboard ).default( [] ),
    currentDashboardId: z.string().optional(),
} )
```

Hope this helps. :D
